server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetMappingAction.java

```diff
@@ -62,13 +71,21 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         getMappingsRequest.indicesOptions(IndicesOptions.fromRequest(request, getMappingsRequest.indicesOptions()));
         getMappingsRequest.masterNodeTimeout(request.paramAsTime("master_timeout", getMappingsRequest.masterNodeTimeout()));
         getMappingsRequest.local(request.paramAsBoolean("local", getMappingsRequest.local()));
-        return channel -> client.admin().indices().getMappings(getMappingsRequest, new RestBuilderListener<>(channel) {
+        return channel -> client.admin().indices().getMappings(getMappingsRequest, new RestActionListener<>(channel) {
+
             @Override
-            public RestResponse buildResponse(final GetMappingsResponse response, final XContentBuilder builder) throws Exception {
-                builder.startObject();
-                response.toXContent(builder, request);
-                builder.endObject();
-                return new BytesRestResponse(RestStatus.OK, builder);
+            protected void processResponse(GetMappingsResponse getMappingsResponse) {
+                // Process serialization on GENERIC pool since the serialization of the raw mappings to XContent can be too slow to execute
+                // on an IO thread
+                threadPool.generic().execute(ActionRunnable.wrap(this, l -> new RestBuilderListener<GetMappingsResponse>(channel) {
+                    @Override
+                    public RestResponse buildResponse(final GetMappingsResponse response, final XContentBuilder builder) throws Exception {
```

Comment:
WDYT about checking for a timeout again here before we do all the serialisation work? If we use a bounded threadpool then these things might pile up, so aborting early might be a helpful way to push back.

Reply:
:+1:  how about https://github.com/elastic/elasticsearch/pull/57937/commits/0bd68d4d6be52152133889a5504bf1ff96874906 ? :)